### PR TITLE
pscanrulesAlpha: Full Path disclosure vulnerability passive scan

### PR DESCRIPTION
**Pull request by ganesh-dagadi**
## Overview
A passive scanner to scan for full path disclosure vulnerability on windows and UNIX systems.

## Related Issues
[Issue link](https://github.com/zaproxy/zaproxy/issues/413)

## Checklist
- [x] Update help
- [x] Update changelog
- [x] Run `./gradlew spotlessApply` for code formatting
- [x] Write tests
- [x] Check code coverage
- [x] Sign-off commits
- [ ] Squash commits
- [x] Use a descriptive title

For more details, please refer to the [developer rules and guidelines](https://www.zaproxy.org/docs/developer/dev-rules-and-guidelines/).

### CODE REVIEW

**Comment by kingthorin** [comment]
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/main/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

**Comment by ganesh-dagadi** [comment]
> To address the DCO requirement you'll need to sign-off the commit(s):
> 
> * https://github.com/zaproxy/zaproxy/blob/main/CONTRIBUTING.md#developer-certificate-of-origin
> * https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

Done

**Comment by thc202** [comment]
Also needs:
> Run `./gradlew spotlessApply` for code formatting

**Comment by ganesh-dagadi** [comment]
I think another commit not related to this PR has popped up after I performed squashing.

File: addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        Pattern pathPattern =
+                Pattern.compile("(?i)(/([a-z0-9]+/)+)|(([a-z0-9]:\\\\)+([a-z0-9]+\\\\)+)");
+        Matcher pathMatcher = pathPattern.matcher(responseBody);
+        if (!(statusCode < 200 || statusCode > 299)) return;
+        if (!(errorMatcher.find())) return;
+        if (!pathMatcher.find()) {
+            evidence = "Found word/s : " + errorMatcher.group();
+            buildAlert(evidence, false).raise();
+        } else {
+            evidence = "Found Possible Full Path/s : " + pathMatcher.group();
```
Comment (kingthorin) [inline addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:59]: Evidence should be only what was actually found in order to be highlighted.

File: addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        Pattern errorPattern = Pattern.compile("(?i)error|parse|fatal|warning|notice");
+        Matcher errorMatcher = errorPattern.matcher(responseBody);
+        // matches the presence of either windows style or unix style full paths
+        // example for windows : C:\folder\folder
+        // example for unix : /dir/dir
+        Pattern pathPattern =
+                Pattern.compile("(?i)(/([a-z0-9]+/)+)|(([a-z0-9]:\\\\)+([a-z0-9]+\\\\)+)");
+        Matcher pathMatcher = pathPattern.matcher(responseBody);
+        if (!(statusCode < 200 || statusCode > 299)) return;
+        if (!(errorMatcher.find())) return;
```
Comment (kingthorin) [inline addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:54]: These should use curly braces for consistency.
Instead of checking status code you should use isPageSuccess. Though I'm not sure I agree with skipping responses for either of these reasons.

File: addOns/pscanrulesAlpha/src/main/resources/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties [hunk 3]
```diff
@@ -18,6 +18,11 @@ pscanalpha.examplefile.name = An example passive scan rule which loads data from
 pscanalpha.examplefile.other = This is for information that doesn't fit in any of the other sections.
 pscanalpha.examplefile.refs = https://www.zaproxy.org/blog/2014-04-03-hacking-zap-3-passive-scan-rules/
 pscanalpha.examplefile.soln = A general description of how to solve the problem.
+pscanalpha.fullpathdisclosurealert.desc = The full path of files which is sensitive has been exposed to the client
+
+pscanalpha.fullpathdisclosurealert.name = Full path disclosure detected
```
Comment (kingthorin) [inline addOns/pscanrulesAlpha/src/main/resources/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties:23]: > Full Path Disclosure

File: addOns/pscanrulesAlpha/src/main/resources/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties [hunk 4]
```diff
@@ -62,4 +67,3 @@ pscanalpha.metadata-request-headers.sfu.missing.refs = https://developer.mozilla
 pscanalpha.metadata-request-headers.sfu.missing.soln = Ensure that Sec-Fetch-User header is included in user initiated requests.
 
 pscanalpha.name = Passive Scan Rules - alpha
-
```
Comment (kingthorin) [inline addOns/pscanrulesAlpha/src/main/resources/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties:65]: Revert this change.

File: addOns/pscanrulesAlpha/src/test/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDiscosureAlertTest.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+    @Test
+    void testSuccessFulResponseWithNoPathDisclosure() {
+        HttpMessage message = new HttpMessage();
+        HttpResponseHeader header = new HttpResponseHeader();
+        header.setStatusCode(200);
+        message.setResponseHeader(header);
+        scanHttpResponseReceive(message);
+        assertThat(alertsRaised, hasSize(0));
+    }
```
Comment (kingthorin) [inline addOns/pscanrulesAlpha/src/test/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDiscosureAlertTest.java:47]:
```suggestion
    @Test
    void shouldNotAlertOnSuccessfulResponseWithNoPathDisclosure() {
        // Given
        HttpMessage message = new HttpMessage();
        HttpResponseHeader header = new HttpResponseHeader();
        header.setStatusCode(200);
        message.setResponseHeader(header);
        // When
        scanHttpResponseReceive(message);
        // Then
        assertThat(alertsRaised, hasSize(0));
    }
```

File: addOns/pscanrulesAlpha/src/test/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDiscosureAlertTest.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                        + "<meta charset=\"utf-8\">\n"
+                        + "<title>Error</title>\n"
+                        + "</head>\n"
+                        + "<body>\n"
+                        + "<pre>Cannot GET /</pre>\n"
+                        + "</body>\n"
+                        + "</html>\n";
+        HttpMessage message = createMessage(testBody, HttpStatusCode.NOT_FOUND);
+        scanHttpResponseReceive(message);
+        if (alertsRaised.size() != 1) fail();
```
Comment (kingthorin) [inline addOns/pscanrulesAlpha/src/test/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDiscosureAlertTest.java:67]: make this an actual assertion. Ex: `assertThat(alertsRaised, hasSize(1))`

File: addOns/pscanrulesAlpha/src/test/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDiscosureAlertTest.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+                        + "</body>\n"
+                        + "</html>\n";
+        HttpMessage msg = createMessage(testBody, HttpStatusCode.NOT_FOUND);
+        scanHttpResponseReceive(msg);
+        if (alertsRaised.size() != 1) fail();
```
Comment (kingthorin) [inline addOns/pscanrulesAlpha/src/test/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDiscosureAlertTest.java:81]: Use an actual assertion.

File: addOns/pscanrulesAlpha/src/test/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDiscosureAlertTest.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
```
Comment (kingthorin) [inline addOns/pscanrulesAlpha/src/test/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDiscosureAlertTest.java:93]: Use an actual assertion.

File: addOns/pscanrulesAlpha/src/test/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDiscosureAlertTest.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        assertEquals(alertsRaised.get(0).getRisk(), Alert.RISK_LOW);
+    }
+
+    @Test
+    void scanReturnsLowConfidenceAlert() throws URIException {
+        HttpMessage msg =
+                createMessage("Error : Cant find /sample/path/", HttpStatusCode.NOT_FOUND);
+        scanHttpResponseReceive(msg);
+        if (alertsRaised.size() != 1) fail();
```
Comment (kingthorin) [inline addOns/pscanrulesAlpha/src/test/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDiscosureAlertTest.java:102]: Same

**Review by kingthorin — COMMENTED** [review]
A help entry should be added.

UnitTests should be named appropriately and all have Given/When/Then comments.
https://www.zaproxy.org/docs/developer/verifying-your-changes/#new-unit-tests

File: addOns/pscanrulesAlpha/CHANGELOG.md [hunk 10]
```diff
@@ -6,6 +6,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ## Unreleased
 ### Changed
 - Update minimum ZAP version to 2.14.0.
+### Added
+- Full Path Disclosure vulnerability passive scanner.
```
Comment (kingthorin) [inline addOns/pscanrulesAlpha/CHANGELOG.md:10]:
```suggestion

### Added
- Full Path Disclosure vulnerability passive scanner.
```

Also add an issue reference (see other entries in the change log.)

File: addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.htmlparser.jericho.Source;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.pscan.PluginPassiveScanner;
+
+public class FullPathDisclosureAlertPassiveScanRule extends PluginPassiveScanner {
+    private final int pluginID = 110009;
```
Comment (kingthorin) [inline addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:31]: Please open a PR against zaproxy/zaproxy scanners.md to reserve an ID.

File: addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java (hunk 2 above)
Comment (ganesh-dagadi) [reply addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:54]: consider this HATEOAS response
`HTTP/1.1 200 OK
Content-Type: application/+json
Content-Length: ...
{
   "payroll": {
       "employee_number": "employee_123",
       "salary" : 1000,
       "links": {
           "increment": "/payroll/employee_123/increment",
           "decrement": "/payroll/employee_123/decrement",
           "close": "/payroll/employee_123/close"
       }
   }
}`
Here `"/payroll/employee_123/increment"` is matched to a unix based file path by regex.
Here is another example that leads to a false positive `<a href=https://example.com/somelink/path >click me</a>`
Thus to avoid false positives, I have introduced these checks.

**Comment by ganesh-dagadi** [comment]
I will do the necessary changes and push. 

> I think another commit not related to this PR has popped up after I performed squashing.

Hope this will not cause a problem.

I'm new to using git on such a large scale. So need some help

1. I will perform my changes and commit them. 
2. I will then do `git rebase -i HEAD~2` and squash the latest commit. I enter 2 as I want to squash the latest commit and the previous one. 
3. After that I will `git push origin main --force`

**Note** : I will perform all these operations on a fork of the zap-extensions repository.

Is there a git workflow you recommend so that I can keep my fork in sync with the main repo without messing up the commits.

File: addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java (hunk 1 above)
Comment (ganesh-dagadi) [reply addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:59]: Should I remove the word 'possible'?

Comment (kingthorin) [reply addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:59]: You should just use the matched group. The literal string that you're concatenating won't be part of the text to match.

File: addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java (hunk 2 above)
Comment (kingthorin) [reply addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:54]: Neither of these checks would address those examples.

It's a 200 so it passes, it doesn't have one of the error words.

Further a redirect 3xx, client error 4xx, or server error 5xx could definitely disclose a full path. In fact IIS is almost guaranteed to on 5xx when custom error pages aren't enabled.

Maybe instead of trying to exclude error words include some default linux and macOS paths?
/bin
/usr
/mnt
/proc
/sbin
/dev
/lib
/tmp
/opt
/home
/var
/root
/etc
(☝ not necessarily ordered in any particularly useful way)

/Applications
/Volumes
/System
/Users
/Developer
/Library
(☝ not necessarily ordered in any particularly useful way)

You can easily add a UnitTest to try :wink:

**Comment by kingthorin** [comment]
Add another commit with good changes, then we can either fix it on your behalf or provide instructions how to fix it.

File: addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java (hunk 11 above)
Comment (ganesh-dagadi) [reply addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:31]: Do I have to include [Script]

Comment (kingthorin) [reply addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:31]: Nope

File: addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java (hunk 2 above)
Comment (ganesh-dagadi) [reply addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:54]: The inclusion of default linux and windows paths in a vulnerable request is a good idea and I will implement it replacing the errorMatcher. However, I am a little confused with the statusCode.

So this response, being a success and not containing any FPD vulnerabilities should not raise any alerts.

```
HTTP/1.1 200 OK 
Content-Type: application/+json 
 Content-Length: ... 
{ "payroll": { "employee_number": "employee_123", "salary" : 1000, "links": { "increment": "/payroll/employee_123/increment", "decrement": "/payroll/employee_123/decrement", "close": "/payroll/employee_123/close" } } }
```

So the line, (taking `statusCode` as `200`)
`if (!(statusCode < 200 || statusCode > 299)) return;` evaluates to `if(!(false || false))` => `if( !(false) )` => `if(true)` will cause `return` to be hit. Thus, no alerts raised.
Only if the response did not succeed (that is 1xx or 3xx or 4xx or 5xx) will it continue to the lines after the guard statement where we check for default vulnerable paths and perform regex pattern matching to extract the found path.

> Neither of these checks would address those examples.

> It's a 200 so it passes, it doesn't have one of the error words.

My understanding contradicts this. Am I missing something here?

I wrote these two unit tests and both of them passed.

```
void sampleTest1() throws URIException{
        //Given
        HttpMessage msg = createMessage("{ \"payroll\": { \"employee_number\": \"employee_123\", \"salary\" : 1000, \"links\": { \"increment\": \"/payroll/employee_123/increment\", \"decrement\": \"/payroll/employee_123/decrement\", \"close\": \"/payroll/employee_123/close\" } } }\n" +
                "Here \"/payroll/employee_123/increment\"" , HttpStatusCode.OK);
        //When
        scanHttpResponseReceive(msg);
        //Then
        assertThat(alertsRaised , hasSize(0));
    }
```

```
void sampleTest2() throws URIException{
        //Given
        HttpMessage msg = createMessage("<a href=\"https://example.com/foo/bar/\">" , HttpStatusCode.OK);
        //When
        scanHttpResponseReceive(msg);
        //Then
        assertThat(alertsRaised , hasSize(0));
    }
```

Comment (kingthorin) [reply addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:54]: That’s my fault, I was ignoring the leading bang/not (!).

Comment (ganesh-dagadi) [reply addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:54]: So I will preserve the statusCheck but make it cleaner by moving the it to a separate private function.

Comment (kingthorin) [reply addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java:54]: No you should use `!isSuccess`

https://github.com/zaproxy/zaproxy/blob/8e78a891d88aeadb6c6007cb2437982eb73536ef/zap/src/main/java/org/parosproxy/paros/core/scanner/AbstractPlugin.java#L747

**Comment by ganesh-dagadi** [comment]
I have pushed a new commit with the necessary changes.

**Comment by kingthorin** [comment]
Thanks, I'll tiddy it up later today and provide instructions of what to do before making more changes.

**Comment by kingthorin** [comment]
Grrr, sorry I botched this.

The branch you want is here: https://github.com/kingthorin/zap-extensions/tree/main-gan I'll provide instructions in a bit.

**Comment by kingthorin** [comment]
Okay here's what you need to do:

```bash
git remote add kingthorin https://github.com/kingthorin/zap-extensions.git
git fetch kingthorin
git checkout kingthorin/main-gan
git push -u origin main-gan
```

Then goto https://github.com/ganesh-dagadi/zap-extensions and it should prompt you to open a PR.
